### PR TITLE
Add new commands: run-cmd, copy-file

### PR DESCRIPTION
**Pull request by fsaintma**
+ Add with_proxy option in ssh lib run command. With with_proxy option you execute the command
on the A8 nodes otherwise on the SSH frontend only at once.

+ Add two parser commands:
run-cmd : execute a command on the A8 nodes or frontend SSH (--frontend option)
copy-file : copy a file on the SSH frontend homedir directory (~/A8/.iotlabsshcli)

+ Modify run-script/flash-m3 command when with_proxy option is needed. (create
directory or change permissions on the SSH frontend)

+ Add --frontend option for run-script command.

### CODE REVIEW

File: iotlabsshcli/open_a8.py [hunk 1]
```diff
@@ -117,14 +119,19 @@ def wait_for_boot(config_ssh, nodes, max_wait=120, verbose=False):
     return {"wait-for-boot": result}
 
 
-def run_cmd(config_ssh, nodes, cmd, verbose=False):
-    """ Run a command on the A8 nodes."""
+def run_cmd(config_ssh, nodes, cmd, frontend, verbose=False):
```
Comment (fambon) [inline iotlabsshcli/open_a8.py]: frontend=None

File: README.md [hunk 2]
```diff
@@ -163,8 +166,7 @@ $ open-a8-cli --verbose run-cmd "uname -a" --frontend
     }
 }
 ```
-* Copy file on SSH frontend homedir directory (eg. ~/A8/.iotlabsshcli). This directory
-is mounting by A8 nodes during experiment.
+#### Copy file on SSH frontend homedir directory (eg. ~/A8/.iotlabsshcli):
```
Comment (fambon) [inline README.md]: This directory is shared via NFS between the ssh frontend and the A8 nodes

File: README.md [hunk 3]
```diff
@@ -183,7 +185,9 @@ $ open-a8-cli run-cmd "tar -xzvf ~/A8/.iotlabsshcli/test.tar.gz -C ~/A8/.iotlabs
     }
 }
 ```
-* Run the script `/tmp/test.sh` on `node-a8-2` in saclay:
+<b>Note:</b> A8 homedir directory is mounting (eg. NFS) by A8 nodes during experiment.
```
Comment (fambon) [inline README.md]: mountED (via NFS) by all A8 nodes during the experiment

File: README.md [hunk 4]
```diff
@@ -194,14 +198,13 @@ $ open-a8-cli run-script /tmp/test.sh -l saclay,a8,2
     }
 }
 ```
-Note: a screen session is launched on the A8 node which
-startup script
+<b>Note:</b> a screen session is launched on the A8 node which startup script
```
Comment (fambon) [inline README.md]: ... on the A8 node to actually run the script and provide easy access to outputs if needed.
When the script ends, the screen session is terminated and the logs are gone.

File: iotlabsshcli/open_a8.py [hunk 5]
```diff
@@ -117,8 +119,55 @@ def wait_for_boot(config_ssh, nodes, max_wait=120, verbose=False):
     return {"wait-for-boot": result}
 
 
-def run_script(config_ssh, nodes, script, verbose=False):
-    """Run a script in background on the A8 nodes."""
+def run_cmd(config_ssh, nodes, cmd, frontend, verbose=False):
```
Comment (fambon) [inline iotlabsshcli/open_a8.py]: frontend=None

Comment (fambon) [inline iotlabsshcli/open_a8.py]: maybe rename parameter to `run_on_frontend` ?

File: iotlabsshcli/open_a8.py [hunk 6]
```diff
@@ -117,8 +119,55 @@ def wait_for_boot(config_ssh, nodes, max_wait=120, verbose=False):
...
+            result = ssh.run(cmd)
+    except OpenA8SshAuthenticationException as exc:
+        print(exc.msg)
+        result = {"1": nodes}
+
+    return {"run-cmd": result}
+
+
+def copy_file(config_ssh, nodes, file_path, verbose=False):
+    """ Copy a file on the A8 SSH frontend directory
```
Comment (fambon) [inline iotlabsshcli/open_a8.py]: on a8 ssh frontend_S_ (there might be several frontends involved in the experiment) ?

File: iotlabsshcli/parser/open_a8_parser.py [hunk 7]
```diff
@@ -47,36 +47,55 @@ def parse_options():
...
     common.add_nodes_selection_list(run_script_parser)
 
+    # run-cmd parser
+    run_cmd_parser = subparsers.add_parser('run-cmd',
+                                           help='Run a command on A8 nodes')
+    run_cmd_parser.add_argument('cmd', help='Command')
+    run_cmd_parser.add_argument('--frontend', action='store_true',
+                                help='Execution on SSH frontend')
     # nodes list or exclude list
-    common.add_nodes_selection_list(boot_parser)
```
Comment (fambon) [inline iotlabsshcli/parser/open_a8_parser.py]: what was this `boot_parser` doing there ?

File: iotlabsshcli/tests/open_a8_parser_test.py [hunk 8]
```diff
@@ -149,7 +157,66 @@ def test_main_run_script(self, list_nodes, run_script):
...
+            open_a8_parser.main(args)
+            list_nodes.assert_called_with(self.api, 123, None, None)
+            run_cmd.assert_called_with({'user': 'username', 'exp_id': 123},
+                                       self._root_nodes,
+                                       'uname -a', False, verbose=False)
+
+    @patch('iotlabsshcli.open_a8.copy_file')
+    @patch('iotlabcli.parser.common.list_nodes')
+    def test_main_copy_file(self, list_nodes, copy_file):
+        """Run the parser.node.main with run-script subparser function."""
```
Comment (fambon) [inline iotlabsshcli/tests/open_a8_parser_test.py]: with the `copy_file sub-parser`

File: README.md [hunk 9]
```diff
@@ -52,7 +54,7 @@ to configure your IoT-LAB SSH access in this
 
 ### Examples:
 
-* Start an experiment, wait for it ot be ready, wait for all A8 boot
+#### Start an experiment, wait for it ot be ready, wait for all A8 boot:
```
Comment (aabadie) [inline README.md]: ot be ready => **to** be ready

The typo was already there but as your are at it, you can change it.

File: README.md [hunk 10]
```diff
@@ -120,28 +139,72 @@ $ open-a8-cli reset-m3 -l saclay,a8,2
...
+$ open-a8-cli run-script /tmp/test.sh -l saclay,a8,2
+{
+    "run-script": {
         "0": [
             "node-a8-2.saclay.iot-lab.info"
         ]
     }
 }
 ```
+<b>Note:</b> a screen session is launched on the A8 node which startup script
```
Comment (aabadie) [inline README.md]: I prefer: the script starts in a screen session launched on each A8 node

File: README.md [hunk 11]
```diff
@@ -120,28 +139,72 @@ $ open-a8-cli reset-m3 -l saclay,a8,2
...
+
+```
+root@node-a8-2:~# screen -ls
+There is a screen on:
+	1877.<login>-<exp_id>	(Detached)
+1 Socket in /tmp/screens/S-root.
+```
+<b>Note:</b> as run command you can pass --frontend option if you want to launch a script
```
Comment (aabadie) [inline README.md]: Similar to run command

File: iotlabsshcli/open_a8.py (hunk 5 above)
Comment (aabadie) [inline iotlabsshcli/open_a8.py]: "backward compat" ?  for a project not released yet ?

File: iotlabsshcli/open_a8.py [hunk 12]
```diff
@@ -117,8 +119,55 @@ def wait_for_boot(config_ssh, nodes, max_wait=120, verbose=False):
...
+def run_cmd(config_ssh, nodes, cmd, frontend, verbose=False):
+    """ Run a command on the A8 nodes or on the
+    SSH frontend
+    """
+
+    # Configure ssh.
+    groups = _nodes_grouped(nodes)
+    ssh = OpenA8Ssh(config_ssh, groups, verbose=verbose)
+    try:
+        if frontend:
```
Comment (aabadie) [inline iotlabsshcli/open_a8.py]: you can use a single line here:
```python
result = ssh.run(cmd, with_proxy=not frontend)
```

File: iotlabsshcli/open_a8.py [hunk 13]
```diff
@@ -141,13 +192,18 @@ def run_script(config_ssh, nodes, script, verbose=False):
         ssh.scp(script, remote_script)
 
         # Make script executable
-        ssh.run(_MAKE_EXECUTABLE_CMD.format(remote_script))
+        ssh.run(_MAKE_EXECUTABLE_CMD.format(remote_script),
+                with_proxy=with_proxy)
+
+        if not frontend:
```
Comment (aabadie) [inline iotlabsshcli/open_a8.py]:
```python
with_proxy = not frontend
```

**Comment by aabadie** [comment]
For the missing coverage, you need to find a way to call `OpenA8Ssh.run` with `with_proxy=False` and do some checks.

Otherwise (I forgot to mention it in my previous review), the changes are important but it's still a good job. Thanks !

File: iotlabsshcli/open_a8.py [hunk 14]
```diff
@@ -117,8 +119,55 @@ def wait_for_boot(config_ssh, nodes, max_wait=120, verbose=False):
     return {"wait-for-boot": result}
 
 
-def run_script(config_ssh, nodes, script, verbose=False):
-    """Run a script in background on the A8 nodes."""
+def run_cmd(config_ssh, nodes, cmd, run_on_frontend=False, verbose=False):
+    """ Run a command on the A8 nodes or on the
```
Comment (aabadie) [inline iotlabsshcli/open_a8.py]: Docstring can be on one line

File: iotlabsshcli/open_a8.py [hunk 15]
```diff
@@ -117,8 +119,55 @@ def wait_for_boot(config_ssh, nodes, max_wait=120, verbose=False):
...
+    SSH frontend
+    """
+
+    # Configure ssh.
+    groups = _nodes_grouped(nodes)
+    ssh = OpenA8Ssh(config_ssh, groups, verbose=verbose)
+    try:
+        if run_on_frontend:
```
Comment (aabadie) [inline iotlabsshcli/open_a8.py]: The `if else` can be rewritten in a single line:
```python
result = ssh.run(cmd, with_proxy=not run_on_frontend)
```

File: iotlabsshcli/sshlib/open_a8_ssh.py [hunk 16]
```diff
@@ -186,12 +187,12 @@ def _try_connection(self, node, site):
         except ConnectionErrorException:
             if self.verbose:
                 print("Node {} not ready."
-                      .format(_node_fqdn(node, site)))
+                      .format(node))
```
Comment (aabadie) [inline iotlabsshcli/sshlib/open_a8_ssh.py]: the print can be rewritten in one line

File: iotlabsshcli/sshlib/open_a8_ssh.py [hunk 17]
```diff
@@ -186,12 +187,12 @@ def _try_connection(self, node, site):
...
         else:
             client.client.close()
             if self.verbose:
                 print("Node {} ready."
-                      .format(_node_fqdn(node, site)))
+                      .format(node))
```
Comment (aabadie) [inline iotlabsshcli/sshlib/open_a8_ssh.py]: same here

**Comment by fsaintma** [comment]
Help please :) It seems that codecov is now inactive on this branch ?
Can we merge this pull request or what is needed ?

**Comment by aabadie** [comment]
@fsaintma, can you:
```
$ git commit --amend
$ git push origin run_cmd -f
```
?
Just to trigger a new build.

**Comment by fambon** [comment]
I'll rebase master + push force.

-------- Original Message --------

**Comment by aabadie** [comment]
> I'll rebase master + push force.

Please don't. It seems that your last pushes are blocking codecov.

**Comment by aabadie** [comment]
for an unknown reason

**Comment by fambon** [comment]
oops, missed the no go.
codecov is apparently undergoing minor service outage http://status.codecov.io/

**Comment by aabadie** [comment]
Thanks for checking. Let's wait then.
In any case, the coverage still misses a few lines.

**Comment by fambon** [comment]
yeah, that's on the ofa_review branch as of now, but on it's way to new_cmd.
how do I check if it's going to better the coverage ? I see no diff using `tox -e tests`

**Comment by aabadie** [comment]
> how do I check if it's going to better the coverage ?

Open a PR ;)
Codevoc merges the coverage of the different builds, that why it's activated for python 2.7 and 3.5.
`tox -e tests` gives a rough overview.

The coverage of this PR can be checked [here](https://codecov.io/gh/iot-lab/ssh-cli-tools/pulls/15). Here are [the missing lines](https://codecov.io/gh/iot-lab/ssh-cli-tools/pulls/15/diff#D4-122).

Side note: try to avoid  cross posting about different branches/PRs. Or maybe I misunderstood.

**Comment by fambon** [comment]
ok, so I saw the same info, but in a different place:
https://codecov.io/gh/iot-lab/ssh-cli-tools/src/new_cmd/iotlabsshcli/sshlib/open_a8_ssh.py#L110

File: tox.ini [hunk 18]
```diff
@@ -16,7 +16,7 @@ commands=
 
 [testenv:tests]
 commands=
-    pytest -v --doctest-modules iotlabsshcli --cov=iotlabsshcli
+    pytest --doctest-modules iotlabsshcli --cov=iotlabsshcli
```
Comment (aabadie) [inline tox.ini]: why removing `-v` ?

File: iotlabsshcli/sshlib/open_a8_ssh.py (hunk 16 above)
Comment (aabadie) [inline iotlabsshcli/sshlib/open_a8_ssh.py]: Weird, I'm pretty sure this comment was addressed.

File: tox.ini (hunk 18 above)
Comment (fambon) [inline tox.ini]: clutters my screen when looking for summary of coverage.  more compact report.  and tests are green anyway :)

**Comment by fambon** [comment]
I merged (--no-ff) the additional tests.

File: tox.ini (hunk 18 above)
Comment (aabadie) [inline tox.ini]: not a good reason sorry. You can't change things like this. This is totally unrelated to this PR !

**Comment by fambon** [comment]
how do you set the coverage threshold in codecov ? I see nothing in `codecov.yml`.  Is it on the server, and I would need an account ?

**Comment by aabadie** [comment]
> how do you set the coverage threshold in codecov ? I see nothing in codecov.yml. Is it on the server, and I would need an account ?

Ask google : https://github.com/codecov/support/wiki/Codecov-Yaml

But I'm against using a threshold. The goal is to keep 100% coverage, except if you have a real good reason.

File: tox.ini (hunk 18 above)
Comment (aabadie) [inline tox.ini]: can you please revert this change ?

File: iotlabsshcli/tests/open_a8_ssh_test.py [hunk 19]
```diff
@@ -29,8 +29,9 @@
 from iotlabsshcli.sshlib.open_a8_ssh import _nodes_from_groups
 from .compat import patch
 
+_SITES = ['saclay', 'grenoble']
```
Comment (aabadie) [inline iotlabsshcli/tests/open_a8_ssh_test.py]: Another unrelated change. Can you please revert it ?

**Comment by aabadie** [comment]
There's a merge commit in this PR. @fsaintma can you rebase ?

**Comment by aabadie** [comment]
Merged ! Thanks @fsaintma and @fambon for the 100% coverage :)
